src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala

```diff
@@ -340,6 +343,62 @@ class LeonardoService(protected val dataprocConfig: DataprocConfig,
+                    clusterRequest: ClusterRequest): Future[Cluster] = {
+    for {
+      cluster <- internalGetActiveClusterDetails(googleProject, clusterName) //throws 404 if nonexistent
+
+      _ <- checkClusterPermission(userInfo, DeleteCluster, cluster) //throws 404 if no auth //TODO: check a more appropriate action here
+
+      _ <- internalUpdateCluster(cluster, clusterRequest)
+
+      updatedCluster <- internalGetActiveClusterDetails(googleProject, clusterName)
+    } yield { updatedCluster }
+  }
+
+  def internalUpdateCluster(existingCluster: Cluster, clusterRequest: ClusterRequest) = {
+    if(existingCluster.status.isUpdatable) {
+      for {
+        _ <- handleUpdatingAutopauseThreshold(existingCluster.id, clusterRequest.autopause, clusterRequest.autopauseThreshold)
+
+        _ <- handleResizingCluster(existingCluster, clusterRequest.machineConfig)
+
+        // Set the cluster status to Updating
+        _ <- dbRef.inTransaction { _.clusterQuery.updateClusterStatus(existingCluster.id, ClusterStatus.Updating) }
+
+        updatedCluster <- internalGetActiveClusterDetails(existingCluster.googleProject, existingCluster.clusterName)
+      } yield {
+        clusterMonitorSupervisor ! ClusterUpdated(updatedCluster.copy(status = ClusterStatus.Updating))
+        updatedCluster
+      }
+    } else Future.failed(ClusterCannotBeUpdatedException(existingCluster.googleProject, existingCluster.clusterName, existingCluster.status))
+  }
+
+  def handleUpdatingAutopauseThreshold(clusterId: Long, autopause: Option[Boolean], autopauseThreshold: Option[Int]): Future[Int] = {
+    (autopause, autopauseThreshold) match {
+      case (None, None) => Future.successful(0) //no-op. throwing None and None into calculateAutopauseThreshold would be wrong so we won't touch it
+      case _ => dbRef.inTransaction { dataAccess => dataAccess.clusterQuery.updateAutopauseThreshold(clusterId, calculateAutopauseThreshold(autopause, autopauseThreshold)) }
+    }
+  }
+
+  def handleResizingCluster(existingCluster: Cluster, machineConfigOpt: Option[MachineConfig]): Future[Unit] = {
+    machineConfigOpt match {
+      case Some(machineConfig) =>
+        for {
+          // Add Dataproc Worker role to the cluster service account, if present.
+          // This is needed to be able to spin up Dataproc clusters.
+          // If the Google Compute default service account is being used, this is not necessary.
+          _ <- addDataprocWorkerRoleToServiceAccount(existingCluster.googleProject, existingCluster.serviceAccountInfo.clusterServiceAccount)
+          _ <- gdDAO.resizeCluster(existingCluster.googleProject, existingCluster.clusterName, machineConfig.numberOfWorkers, machineConfig.numberOfPreemptibleWorkers)
+          _ <- dbRef.inTransaction { dataAccess => machineConfig.numberOfWorkers.map(numWorkers => dataAccess.clusterQuery.updateNumberOfWorkers(existingCluster.id, numWorkers)) getOrElse DBIO.successful(0) }
```

Comment:
I'm wondering it make sense to invert this, to avoid even starting a DB transaction if we don't need to.
```
          _ <- machineConfig.numberOfWorkers.map { numWorkers =>
            dbRef.inTransaction { _.clusterQuery.updateNumberOfWorkers(existingCluster.id, numWorkers) }
          } getOrElse Future.successful(0)
```